### PR TITLE
void object fields are now ignored by codegen and fields/fieldPairs iterator

### PR DESCRIPTION
**Pull request by nc-x**
Closes https://github.com/nim-lang/Nim/issues/3734

### CODE REVIEW

**Comment by nc-x** [comment]
Fails due to this workaround https://github.com/nim-lang/Nim/commit/1fb5dd247745e06c85a2ed1ad8f8f09f0a38b9b9

**Comment by Araq** [comment]
The ``void`` type for object fields is allowed though and the codegens should ignore these fields. I think.

**Comment by nc-x** [comment]
Why should they be allowed then? 
Also lets/vars/consts are not allowed to have type void, so why object fields?

**Comment by nc-x** [comment]
cc @Araq

**Comment by Araq** [comment]
Thanks but your ``system.nim`` made me reconsider. It's yet another special case that gets exposed to every piece of code that uses field iteration. --> Better not allow it and require code to use the more explicit ``when T isnot void`` variant for the outlined use case "Table[K, void] is Set[K]".

**Comment by nc-x** [comment]
> Alternatively the fields iterators should ignore void fields automatically.
  
Done.

File: compiler/semfields.nim [hunk 1]
```diff
@@ -19,6 +19,9 @@ type
     c: PContext
 
 proc instFieldLoopBody(c: TFieldInstCtx, n: PNode, forLoop: PNode): PNode =
+  if c.tupleType == nil and c.field.typ.kind == tyVoid:
```
Comment (Araq) [inline compiler/semfields.nim]: Why the check for ``c.tupleType == nil``?

**Comment by nc-x** [comment]
cc @Araq

**Comment by krux02** [comment]
interesting that void object fields are even allowed. But apart from that, I am ok with these changes.

**Comment by Araq** [comment]
You should use ``isEmptyType`` instead of testing for ``tyVoid`` directly.

**Comment by nc-x** [comment]
Done.

**Pull request merged**